### ISSUE TITLE
Cannot build release 0.7 API docs for knative/eventing-contrib

### ISSUE THREAD

**Issue by RichieEscarez** [knative/docs#1528]
Following the [instructions to build the Knative API docs](https://github.com/knative/docs/blob/master/docs/reference/README.md) fails (see error below) for the knative/eventing-contrib repo.

The [gen-api-reference-docs.sh](https://github.com/knative/docs/blob/master/hack/gen-api-reference-docs.sh) script looks for the API in the [repo]`/pkgs/apis/` folder but that is no longer the structure of the knative/eventing-contrib repo, [for example](https://github.com/knative/eventing-contrib/tree/master/contrib/awssqs/pkg/apis).

Error:
```
I0625 22:01:32.404033   65566 main.go:118] parsing go packages in directory ./pkg/apis
W0625 22:01:32.404871   65566 parse.go:224] Ignoring directory ./pkg/apis: unable to import "./pkg/apis": cannot find package "./pkg/apis" in:
	/var/folders/gq/__83tbn10y753yg0kmgnhcyr009dwx/T/tmp.sc2JgEAe/src/github.com/knative/eventing-contrib/pkg/apis
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14096e5]

goroutine 1 [running]:
k8s.io/gengo/parser.(*Builder).AddDirRecursive(0xc0001a2000, 0x7ffeefbff9bc, 0xa, 0x0, 0x0)
	/Users/rescarez/go/pkg/mod/k8s.io/gengo@v0.0.0-20190327210449-e17681d19d3a/parser/parse.go:229 +0xb5
main.parseAPIPackages(0x7ffeefbff9bc, 0xa, 0x151bc04, 0x23, 0xc00014ff08, 0x1, 0x1)
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.0.0-20190107211124-5c208a662b46/main.go:186 +0x60
main.main()
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.0.0-20190107211124-5c208a662b46/main.go:119 +0x222
```

cc/ @ahmetb

**Comment by RichieEscarez** [comment]
For 0.6 and earlier, this is the page that gets generated and published on knative.dev: https://knative.dev/docs/reference/eventing/eventing-sources-api/

Now that we have "knative/eventing-contrib", do we still need to build API ref docs from the repo?

**Comment by n3wscott** [comment]
Yes and there are sources in `knative/eventing` too

**Comment by ahmetb** [comment]
Looks like there’s a bug in the gengo/parser we’re using. I will try to find some time to debug it.

**Comment by RichieEscarez** [comment]
> Yes and there are sources in `knative/eventing` too

@n3wscott, to clarify, is the following folder where the API docs should be generated for sources/channels (in knative/eventing):
https://github.com/knative/eventing/tree/master/contrib

**Comment by RichieEscarez** [comment]
> > Yes and there are sources in `knative/eventing` too
> 
> @n3wscott, to clarify, is the following folder where the API docs should be generated for sources/channels (in knative/eventing):
> https://github.com/knative/eventing/tree/master/contrib

Thanks @ahmetb! Here are more details from a meeting with Scott:

UPDATE:

Re: sources in `knative/eventing` 
- The API content in `knative/eventing` already gets picked up by the existing API doc gen tool. The built output is staged here (from `master`): https://knative.dev/development/reference/eventing/#sources.eventing.knative.dev

**TODO**

Generate API ref docs for the available ["eventing resources that conform to Knative Eventing"](https://github.com/knative/eventing-contrib/blob/release-0.7/hack/update-codegen.sh#L28) that live in `knative/eventing-contrib` (currently 5 'resources', all in different folders):

```
API_DIRS=(contrib/github/pkg contrib/gcppubsub/pkg camel/source/pkg kafka/source/pkg contrib/awssqs/pkg)
```

Note: there is not a clear pattern nor naming convention, and this list is expected to grow/change

**Comment by ahmetb** [comment]
@RichieEscarez can you share the full command you're running? Since I'm not following the OSS repos (e.g. I didn't add `KNATIVE_EVENTING_SOURCES_REPO="github.com/knative/eventing-contrib"`) I don't know much about what's going on here.

**Comment by ahmetb** [comment]
@n3wscott @RichieEscarez some thoughts below:

tl;dr unless eventing-contrib repo adheres some standard directory structure in the long-term, supporting that repo's api docs generation will not be easy (i.e. I personally wouldn't), and it would cause more issues like this.

1. `-contrib` repo sounds like unofficial stuff, I'm thinking maybe it's not P1 to have ref docs for these.

2. `-contrib` repo HAS TO follow the conventions that the generator tool relies upon (such as directory path naming like `./pkg/apis/xxx/v1alpha1/`, `_types.go` and various meta-comments in .go files like `//+apiGroup`)

3. currently I'm guessing it's failing because we invoke the tool on each repo's `./pkg/apis` directory relative to `/` root, but `events-contrib` has no such thing:

    ```sh
        env GOPATH="${gopath}" "${refdocs_bin}" \
            -out-file "${out_dir}/${out_file}" \
            -api-dir "./pkg/apis" \
    ````

    Maybe we need to refactor this `gen_refdocs` shell function to not to hardcode `-api-dir`, but this means we need to call it MANY times from `main()` function, and it would probably cause having multiple files like `kafka.md`, `pubsub.md`, `camel.md` so our Tech Writers need to actively link to those files.

Happy to discuss further.

**Comment by ahmetb** [comment]
I have a patch in https://github.com/knative/docs/pull/1552

The generator script wasn't working for two reasons:


1. We used `-api-dir ./pkg/apis` for all repos, now it's specified per-repo.

2. When we specified `-api-dir=.`, the generator tool scanned for all packages in the repo (which is what we want) but a package from `vendor/` was picked up https://github.com/knative/eventing-contrib/tree/master/vendor/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/ and it has a _types.go file that's compatible with the generator. (but the directory name is not `vN{alpha|beta}M`) so the generator was failing. &rarr; I released v0.1.3 of the generator tool to ignore packages coming from `*/vendor/*`.

This should be good to go now. Please re-generate with this patch and see if anything is broken with Serving Reference docs with the change I made ignoring `/vendor/` (hopefully not).
